independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java

```diff
@@ -202,6 +207,36 @@ public List<RemovedBean> get() {
         this.contexts = contextsBuilder.build();
     }
 
+    static void precomputeBeanRawTypes(Map<String, List<InjectableBean<?>>> map, InjectableBean<?> bean) {
+        for (Type type : bean.getTypes()) {
+            if (Object.class.equals(type)) {
+                continue;
+            }
+            Class<?> rawType = Types.getRawType(type);
+            if (rawType == null) {
+                continue;
+            }
+            rawType = Types.boxedClass(rawType);
+            String key = rawType.getName();
+            List<InjectableBean<?>> match = map.get(key);
+            if (match == null) {
+                map.put(key, List.of(bean));
+            } else {
+                if (match.contains(bean)) {
+                    continue;
+                }
+                if (match.size() == 1) {
+                    List<InjectableBean<?>> newMatch = new ArrayList<>();
```

Comment:
make it sized as 2: despite we can have several ways to enlarge it; we would like to "not" keep on enlarging it.
Meaning we can make 2 passes 2:
- one with arraylists (not sized)
- second one creating arrays out of them (to have a more compacted version)

The startup will create some more garbage but at runtime you'll have the most compacted version

Reply:
> This make me think that we can perform 2 passes:

Or simply:
```
beansByRawType.forEach(new BiConsumer<String, List<InjectableBean<?>>>() {
            @Override
            public void accept(String key, List<InjectableBean<?>> val) {
                if (val.size() > 1) {
                    ((ArrayList<InjectableBean<?>>)val).trimToSize();
                }
            }
        });
```

> It has to be measured in a microbenchs, but the second one can be a ClassValue<InjectableBean[]> too, that won't require any hashing/collision resolution etc etc

`ClassValue` cannot be used here easily because it's not possible to compute the value from the key (a raw type of a bean type).

> It has to be measured in a microbenchs...

So far I only measured the throughtput (which is like 100x better for a larger app). But I will try to measure the RSS via @Ladicek's tool (https://github.com/Ladicek/arc-crazybeans) asynchronously (i.e. not bound to this PR, so that we can update the logic afterwards if needed).

Reply:
Good point, I have totally forgot about such method of array list, although, if we can use arrays, it would be even better anyway. No polimorphism, no inlining depth worries and no worries of additional small mem footprint costs

Reply:
Good point, I have totally forgot about such method of array list, although, if we can use arrays, it would be even better anyway. No polimorphism, no inlining depth worries and no worries of additional small mem footprint costs.

Said that I am still happy with the current solution eh

Reply:
I think even now it's much better than what we had so unless it's absolutely critical perf bottleneck (which shouldn't be the case here), I'd prefer collections over plain arrays ;-)